x/staking/keeper/slash.go

```diff
@@ -414,6 +414,29 @@ func (k Keeper) SlashRedelegation(ctx context.Context, srcValidator types.Valida
 		}
 	}
 
+	delegation, err := k.GetDelegation(ctx, delegatorAddress, valDstAddr)
+	if err != nil {
+		return math.ZeroInt(), err
+	}
+
+	// if the delegator holds a validator bond to destination validator, decrease the destination validator bond shares
+	if delegation.ValidatorBond {
+		if err := k.SafelyDecreaseValidatorBond(ctx, valDstAddr, math.LegacyDec(totalSlashAmount)); err != nil {
```

Comment:
should we decrease bonded amount by totalSlashAmount or by total sharesToUnbond?

Reply:
I think we should move all this code within the for loop after calling the Unbond method. Basically, decrease the validator bond and both the global and validator liquid totals after every call to Unbond. This will avoid weird behavior when not the entire totalSlashAmount was unbonded. Also, calls to GetDelegation should not return an error (see the call inside the for loop).

Reply:
moved code in the loop and added invariant for liquid stake (for bonded tokens already existed) [b6edd57](https://github.com/cosmos/cosmos-sdk/pull/22538/commits/b6edd5739a6c6d9aac23f508219d96d087eadf37)

Reply:
> for bonded tokens already existed

You mean there was already an invariant for validator bond delegations? If so, how come it was not triggered by this issue?

Reply:
Could you also add an invariant for the validator liquid shares?

Reply:
yeah, i've added for validator liquid shares. Will check the triggering for the bond delegations to see what's happening

Reply:
i improved TestSlashRedelegation because it didnt test slashing of redelegated bonded fund and self-bonded funds, it tested only unbonded redelegations. Also i've changed existing invariant for validator delegations because it didnt checked val bonded delegations just not bonded. Now the invariant is triggered and checked. I also confirmed that with the old code it did not worked.  
TestSlashRedelegation should also contain case for liquid shares, will add that as well and then the pr will be finalized